pkg/apis/kf/v1alpha1/app_types_test.go

```diff
@@ -0,0 +1,143 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"testing"
+
+	"github.com/google/kf/pkg/kf/testutil"
+	"github.com/knative/serving/pkg/apis/autoscaling"
+)
+
+func intPtr(val int) *int {
+	return &val
+}
+
+func TestAppSpecInstances_MinAnnotationValue(t *testing.T) {
+	cases := map[string]struct {
+		instances AppSpecInstances
+		expected  string
+	}{
+		"stopped": {
+			instances: AppSpecInstances{Stopped: true},
+			expected:  "0",
+		},
+		"min defined": {
+			instances: AppSpecInstances{Min: intPtr(3)},
+			expected:  "3",
+		},
+		"exactly defined": {
+			instances: AppSpecInstances{Exactly: intPtr(33)},
+			expected:  "33",
+		},
+		"empty": {
+			instances: AppSpecInstances{},
+			expected:  "",
+		},
+	}
+
+	for tn, tc := range cases {
+		t.Run(tn, func(t *testing.T) {
+			actual := tc.instances.MinAnnotationValue()
+
+			testutil.AssertEqual(t, "annotation", tc.expected, actual)
+		})
+	}
+}
+
+func TestAppSpecInstances_MaxAnnotationValue(t *testing.T) {
+	cases := map[string]struct {
+		instances AppSpecInstances
+		expected  string
+	}{
+		"stopped": {
+			instances: AppSpecInstances{Stopped: true},
+			expected:  "0",
+		},
+		"max defined": {
+			instances: AppSpecInstances{Max: intPtr(3)},
+			expected:  "3",
+		},
+		"exactly defined": {
+			instances: AppSpecInstances{Exactly: intPtr(33)},
+			expected:  "33",
+		},
+		"empty": {
+			instances: AppSpecInstances{},
+			expected:  "",
+		},
+	}
+
+	for tn, tc := range cases {
+		t.Run(tn, func(t *testing.T) {
+			actual := tc.instances.MaxAnnotationValue()
+
+			testutil.AssertEqual(t, "annotation", tc.expected, actual)
+		})
+	}
+}
+
+func TestAppSpecInstances_ScalingAnnotations(t *testing.T) {
+	cases := map[string]struct {
+		instances AppSpecInstances
+		expected  map[string]string
+	}{
+		"stopped": {
+			instances: AppSpecInstances{Stopped: true, Exactly: intPtr(10)},
+			expected: map[string]string{
+				autoscaling.MinScaleAnnotationKey: "0",
+				autoscaling.MaxScaleAnnotationKey: "0",
+			},
+		},
+		"max defined": {
+			instances: AppSpecInstances{Max: intPtr(30)},
+			expected: map[string]string{
+				autoscaling.MaxScaleAnnotationKey: "30",
+			},
+		},
+		"min defined": {
+			instances: AppSpecInstances{Min: intPtr(3)},
+			expected: map[string]string{
+				autoscaling.MinScaleAnnotationKey: "3",
+			},
+		},
+		"range defined": {
+			instances: AppSpecInstances{Min: intPtr(3), Max: intPtr(5)},
+			expected: map[string]string{
+				autoscaling.MinScaleAnnotationKey: "3",
+				autoscaling.MaxScaleAnnotationKey: "5",
+			},
+		},
+		"exactly defined": {
```

Comment:
Can you add a test where exactly min and max are all set, and the test shows that exactly has higher preference?